draft-ietf-ppm-dap.md

```diff
@@ -2117,29 +2110,43 @@ Helper knows it can clean up its state.
-aggregation. In particular, the intent is to allow recovery from a scenario
-where the Helper successfully advances from step `n` to `n+1`, but its
-`AggregationJobResp` response to the Leader gets dropped due to something like
-a transient network failure. The Leader could then resend the request to have
-the Helper advance to step `n+1` and the Helper should be able to retransmit
-the `AggregationJobResp` that was previously dropped. To make that kind of
-recovery possible, Aggregator implementations SHOULD checkpoint the most recent
-step's prep state and messages to durable storage such that the Leader can
-re-construct continuation requests and the Helper can re-construct continuation
-responses as needed.
-
-When implementing an aggregation step skew recovery strategy, the Helper SHOULD
-ensure that the Leader's `AggregationJobContinueReq` message did not change when
-it was re-sent (i.e., the two messages must be identical). This prevents the
-Leader from re-winding an aggregation job and re-running an aggregation step
-with different parameters.
-
-One way the Helper could address this would be to store a digest of the Leader's
-request, indexed by aggregation job ID and step, and refuse to service a request
-for a given aggregation step unless it matches the previously seen request (if
-any).
+Aggregators to ensure that their peer is on the expected step of DAP
+aggregation. An aggregation job is said to be "skewed" if the Leader and Helper
+have a different view of the last completed step of aggregation; for example, if
+the Leader does not receive or successfully process an `AggregationJobResp` from
+the Helper, the Leader will believe that the last completed step is one less
+than the Helper. The intent of the `step` field is to allow detection and
+recovery from a skewed aggregation job. When the Leader receives a
+`stepMismatch` error, it SHOULD re-send the original `AggregationJobInitReq`
+message to restart the aggregation process and allow aggregation to complete
+successfully. To make this kind of recovery possible, Leader implementations
+SHOULD retain enough information to reconstruct the `AggregationJobInitReq`
+message in the case that the aggregation process becomes desynchronized.
+
+Changing an aggregation job's parameters is illegal, so the Helper MUST ensure
+that the Leader's `AggregationJobInitReq` message did not change when it was
+re-sent (i.e., the two messages must be identical). This prevents the Leader
+from re-winding an aggregation job and re-running aggregation with different
+parameters.
+
+One way the Helper could implement this would be to store a digest of the
+Leader's request, indexed by aggregation job ID, and refuse to service a re-sent
+request unless it matches the previously seen request.
+
+Implementation note: in the case that the Leader restarts an aggregation which
+the Helper considers to be complete, the Helper must ensure it does not commit
```

Comment:
```suggestion
the Helper considers to be complete, the Helper MUST NOT commit
```

Reply:
Done. Is it appropriate to include a `MUST` in an implementation note? (If not, I think this behavior is important enough to promote from an implementation note into regular text.)

Reply:
Yes, I think it's appropriate.